### ISSUE TITLE
Symbolic link without actual file link caused FileNotFoundException

### ISSUE THREAD

**Issue by tingung** [zeroturnaround/zt-zip#122]
At https://github.com/zeroturnaround/zt-zip/blob/master/src/main/java/org/zeroturnaround/zip/ZipUtil.java#L1736 - 
String[] filenames = dir.list(); doesn't guarantee return the list of files which are exists in the directory. If there is a symbolic link which link to not exists file, dir.list will return the file name but access the file will throw FileNotFoundException.

Steps to reproduce (must at Linux environment)
1. touch file
2. ln -s file link
3. run ZipUtil::pack to pack the directory contains file create above
4. rm file
5. observe the exception. FileNotFoundException

The solution should always ensure the file is exists. 

`
    for (int i = 0; i < filenames.length; i++) {
      String filename = filenames[i];
      File file = new File(dir, filename)
      if(!file.exists()) {
          continue;
      }
`

**Comment by toomasr** [comment]
I took a look and I'm not sure what is the best solution here. You are trying to pack a folder with broken links, right? Best result is that the links itself end up in the archive and just continue being broken?

This is very similar functionality to `zip --symlink` option of the zip utility. The solution that you provided will skip the broken links though and not sure if optimal.

We use Apache Commons for copying but since Java 7 there is the Files.copy method https://docs.oracle.com/javase/7/docs/api/java/nio/file/Files.html#copy(java.io.InputStream,%20java.nio.file.Path,%20java.nio.file.CopyOption...) which actually will copy the symbolic links and not the destination.

Currently we are supporting Java 6 but not sure preserving this is a great idea.

I think the most correct solution would be update to Java 7 and start preserving the links.

**Comment by tingung** [comment]
I am not sure if the CopyOption is actually can resolve the issue because it is only able to support either follow the symlink or not follow. https://docs.oracle.com/javase/7/docs/api/java/nio/file/LinkOption.html.

For this issue, we should either preserving the links as it is or discard the symlink if preserving is not possible.

**Comment by toomasr** [comment]
Quick test on my Mojave showed that it works in a way where it copies over the link - whether broken or not. I see that in alter versions of Commons IO it also relies on this https://docs.oracle.com/javase/7/docs/api/java/nio/file/StandardCopyOption.html#REPLACE_EXISTING

**Comment by tingung** [comment]
o...okay. great. I think I used the wrong option (LinkOption).

**Comment by toomasr** [comment]
I think I'll throw away the current copying, start relying on Java 7 and bump the requirements. Anybody still on Java 6 will have to do with the current functionality.

**Comment by goxr3plus** [comment]
Yes let's throw Java 6 and move on Java 7 .  Very few companies are still using Java 6 .
